models/work-item-link-repository.go

```diff
@@ -0,0 +1,180 @@
+package models
+
+import (
+	"fmt"
+	"log"
+	"strconv"
+
+	"golang.org/x/net/context"
+
+	"github.com/almighty/almighty-core/app"
+	"github.com/jinzhu/gorm"
+	satoriuuid "github.com/satori/go.uuid"
+)
+
+const (
+	EndpointWorkItemTypes          = "workitemtypes"
+	EndpointWorkItems              = "workitems"
+	EndpointWorkItemLinkCategories = "workitemlinkcategories"
+	EndpointWorkItemLinkTypes      = "workitemlinktypes"
+	EndpointWorkItemLinks          = "workitemlinks"
+)
+
+// NewWorkItemLinkRepository creates a work item link repository based on gorm
+func NewWorkItemLinkRepository(db *gorm.DB) *GormWorkItemLinkRepository {
+	return &GormWorkItemLinkRepository{db}
+}
+
+// GormWorkItemLinkRepository implements WorkItemLinkRepository using gorm
+type GormWorkItemLinkRepository struct {
+	db *gorm.DB
+}
+
+// Create creates a new work item link in the repository.
+// Returns BadParameterError, ConversionError or InternalError
+func (r *GormWorkItemLinkRepository) Create(ctx context.Context, link *WorkItemLink) (*app.WorkItemLink, error) {
+	if err := link.CheckValidForCreation(); err != nil {
+		return nil, err
+	}
+
+	// Fetch the work item link type first in order to check that
+	// the given source and target work items match the correct work item types.
+	linkType := WorkItemLinkType{}
+	db := r.db.Where("id=?", link.LinkTypeID).Find(&linkType)
+	if db.RecordNotFound() {
+		return nil, NewBadParameterError("work item link type", link.LinkTypeID.String())
+	} else if db.Error != nil {
+		fmt.Printf("\n\nError: %s\n\n", db.Error.Error())
+		return nil, NewInternalError(fmt.Sprintf("Failed to find work item link type: %s", db.Error.Error()))
+	}
+
+	// Fetch the source work item
+	source := WorkItem{}
+	db = r.db.Where("id=?", link.SourceID).Find(&source)
+	if db.RecordNotFound() {
+		return nil, NewBadParameterError("source work item", strconv.FormatUint(link.SourceID, 10))
+	} else if db.Error != nil {
+		return nil, NewInternalError(fmt.Sprintf("Failed to find source work item: %s", db.Error.Error()))
+	}
+
+	// Fetch the target work item
+	target := WorkItem{}
+	db = r.db.Where("id=?", link.TargetID).Find(&target)
+	if db.RecordNotFound() {
+		return nil, NewBadParameterError("target work item", strconv.FormatUint(link.TargetID, 10))
+	} else if db.Error != nil {
+		return nil, NewInternalError(fmt.Sprintf("Failed to find target work item: %s", db.Error.Error()))
+	}
+
+	// Check that the work item types of the source and target match those specified in the link type
+	if source.Type != linkType.SourceTypeName {
```

Comment:
Please open an issue for this once we have a basic work item linking in place. This PR is getting way to complex even without this.